### PR TITLE
ARROW-16539: [C++] Bump bundled thrift to 0.16.0

### PR DESCRIPTION
**Pull request by nealrichardson**

### CODE REVIEW

**Comment by nealrichardson** [comment]
Appveyor failure looks like something with gtest, so unrelated. The other Windows failure is https://github.com/apache/arrow/runs/6393577373?check_suite_focus=true#step:7:1065, not sure what that's about. https://github.com/apache/thrift/blame/master/build/cmake/DefinePlatformSpecifc.cmake#L67 mentions this lib, and https://github.com/apache/thrift/commit/7f9abb1cc0f4b2793a48f45ddfcf0d2b287cc50c is the only change to this file since 0.13, so I'm guessing the issue is elsewhere.

If we can resolve this and want to move forward, we should of course run all of the nightlies to check for more issues.

File: cpp/cmake_modules/ThirdpartyToolchain.cmake [hunk 1]
```diff
@@ -1348,21 +1348,22 @@ macro(build_thrift)
...
-      -DWITH_HASKELL=OFF
-      -DWITH_CPP=ON
-      -DWITH_STATIC_LIB=ON
+      -DWITH_JAVASCRIPT=OFF
       -DWITH_LIBEVENT=OFF
-      # Work around https://gitlab.kitware.com/cmake/cmake/issues/18865
-      -DBoost_NO_BOOST_CMAKE=ON)
+      -DWITH_NODEJS=OFF
+      -DWITH_PYTHON=OFF
+      -DWITH_STATIC_LIB=ON)
```
Comment (kou) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake:1366]: Could you also add `-DWITH_ZLIB=OFF`?

**Review by kou — COMMENTED** [review]
> Appveyor failure looks like something with gtest, so unrelated.

Yes. #13101 will fix this.

> The other Windows failure is https://github.com/apache/arrow/runs/6393577373?check_suite_focus=true#step:7:1065, not sure what that's about. https://github.com/apache/thrift/blame/master/build/cmake/DefinePlatformSpecifc.cmake#L67 mentions this lib, and [apache/thrift@7f9abb1](https://github.com/apache/thrift/commit/7f9abb1cc0f4b2793a48f45ddfcf0d2b287cc50c) is the only change to this file since 0.13, so I'm guessing the issue is elsewhere.

https://github.com/apache/thrift/commit/66d110bcce5c7e1b0deeacd28b046fb4a9c0249b /  https://github.com/apache/thrift/pull/2018 may be a related change.

How about force disabling the debug postfix?

```diff
diff --git a/cpp/cmake_modules/ThirdpartyToolchain.cmake b/cpp/cmake_modules/ThirdpartyToolchain.cmake
index 48a0126743..7f85180464 100644
--- a/cpp/cmake_modules/ThirdpartyToolchain.cmake
+++ b/cpp/cmake_modules/ThirdpartyToolchain.cmake
@@ -1355,6 +1355,7 @@ macro(build_thrift)
       -DBUILD_SHARED_LIBS=OFF
       -DBUILD_TESTING=OFF
       -DBUILD_TUTORIALS=OFF
+      -DCMAKE_DEBUG_POSTFIX=
       -DWITH_AS3=OFF
       -DWITH_CPP=ON
       -DWITH_C_GLIB=OFF
@@ -1383,9 +1384,6 @@ macro(build_thrift)
       list(APPEND THRIFT_CMAKE_ARGS "-DWITH_MT=OFF")
     endif()
   endif()
-  if(${UPPERCASE_BUILD_TYPE} STREQUAL "DEBUG")
-    set(THRIFT_STATIC_LIB_NAME "${THRIFT_STATIC_LIB_NAME}d")
-  endif()
   set(THRIFT_STATIC_LIB
       "${THRIFT_PREFIX}/lib/${THRIFT_STATIC_LIB_NAME}${CMAKE_STATIC_LIBRARY_SUFFIX}")
```

File: cpp/cmake_modules/ThirdpartyToolchain.cmake (hunk 1 above)
Comment (nealrichardson) [reply cpp/cmake_modules/ThirdpartyToolchain.cmake:1366]: Sure, sorry I missed those

**Comment by nealrichardson** [comment]
Still no luck @kou. I've tried a few more things. I'm not sure how much this is worth pushing on, I don't think there's any great benefit to upgrading other than generally trying to stay up to date/amortize the cost of updating if there is a critical fix in the future.

**Comment by kou** [comment]
I'll try something in this branch.

**Comment by kou** [comment]
CI is green.
This is ready to merge.

**Comment by nealrichardson** [comment]
@github-actions crossbow submit -g nightly

**Comment by nealrichardson** [comment]
Thanks @kou!

**Pull request closed without merging**